contributingGuides/TS_STYLE.md

```diff
@@ -0,0 +1,335 @@
+
+  ```ts
+  type Foo = {
+    bar: string;
+  };
+
+  // good
+  type ReadOnlyFoo = Readonly<Foo>;
+
+  // bad
+  type ReadOnlyFoo = {
+    readonly [Property in keyof Foo]: Foo[Property];
+  };
+  ```
+
+<a name="convension-object-type"></a><a name="1.13"></a>
+
+- [1.13](#convension-object-type) **`object`**: Don't use `object` type.
+
+> Why? `object` refers to "any non-primitive type," not "any object". Typing "any non-primitive value" is not commonly needed.
+
+```ts
+// bad
+const foo: object = [1, 2, 3]; // TypeScript does not error
+```
+
+If you know that the type of data is an object but don't know what properties or values it has beforehand, use `Record<string, unknown>`.
+
+<a name="convension-export-prop-types"></a><a name="1.14"></a>
+
+- [1.14](#convension-export-prop-types) **Prop Types**: Define and export prop types for components. Use exported prop types instead of grabbing the prop type from a component.
+
+> Why? Exporting prop types aids reusability.
+
+```tsx
+// MyComponent.tsx
+export type MyComponentProps = {
+  foo: string;
+};
+
+export default function MyComponent({ foo }: MyComponentProps) {
+  return <Text>{foo}</Text>;
+}
+
+// bad
+import { ComponentProps } from "React";
+import MyComponent from "./MyComponent";
+type MyComponentProps = ComponentProps<typeof MyComponent>;
+
+// good
+import MyComponent, { MyComponentProps } from "./MyComponent";
+```
+
+## Communication Items
+
+> Comment in the `#expensify-open-source` Slack channel if any of the following situations are encountered. Each comment should be prefixed with `TS ATTENTION:`. Internal engineers will access each situation and prescribe solutions to each case. Internal engineers should refer to general solutions to each situation that follows each list item.
+
+- I think types definitions in a third party library is incomplete or incorrect
+
+  When the library indeed contains incorrect type definitions and it cannot be updated, use module argumentation to correct them.
+
+  ```ts
+  declare module "external-library-name" {
+    interface LibraryComponentProps {
+      // Add or modify typings
+      additionalProp: string;
+    }
+  }
+  ```
+
+## Migration Guidelines
+
+> This section contains instructions that are applicable during the migration.
+
+- Found type bugs. Now what?
+
+  If TypeScript migration uncovers a bug that has been “invisible,” there are two options an author of a migration PR can take
+
+  - Fix issues if they are minor. Document each fix in the PR comment
+  - Suppress a TypeScript error stemming from the bug with `@ts-expect-error`. Create a separate GH issue. Prefix the issue title with `[TS ERROR #<issue-number-of-migration-PR>]`. Cross-link the migration PR and the created GH issue
+
+  The `@ts-expect-error` annotation tells the TS compiler to ignore any errors in the line that follows it. However, if there's no error in the line, TypeScript will also raise an error.
+
+  ```ts
+  // @ts-expect-error
```

Comment:
We should always include a comment along with `@ts-expect-error` that explains why the annotation is used.

Reply:
should we add `TODO` comment?

in [the original PR,](https://github.com/Expensify/App/pull/21050/files?short_path=a7d9fcd#diff-a7d9fcd14fc9df0f15f28cecd203ac29b6b6c8ece970e5015488e99e8172805d:~:text=Usage%20of%20%40ts%2Dexpect%2Derror) there was no comment for `@ts-expect-error` in the example

Reply:
I remember we discussed that comments have to follow `@ts-ignore`, but for `@ts-expect-error` the example in the original PR also didn't include any comment.

nonetheless, I think requiring `TODO` comment with the GH issue number (an issue is always created for every @ts-expect-error`) is a good idea

I updated the section accordingly

Reply:
> in [the original PR,](https://github.com/Expensify/App/pull/21050/files?short_path=a7d9fcd#diff-a7d9fcd14fc9df0f15f28cecd203ac29b6b6c8ece970e5015488e99e8172805d:~:text=Usage%20of%20%40ts%2Dexpect%2Derror) there was no comment for @ts-expect-error in the example

This must have been a mistake. Our ESLint requires a description in the same line:
![image](https://github.com/Expensify/App/assets/46095609/3d165aa1-cd3c-44b1-ab57-aeee348f0da8)

So we should add a description or TODO with issue number like this:
```ts
// @ts-expect-error TODO: #21647
const x: number = 'This is a string'; // No TS error raised
```

Reply:
TODO comment has to be in the same line, without it ESLint errors @hayata-suenaga.

Reply:
changed